binding/Binding/HandleDictionary.cs

```diff
@@ -212,4 +213,115 @@ internal static void DeregisterHandle (IntPtr handle, SKObject instance)
 			}
 		}
 	}
+
+
+	/*
+	 * This is a (hopefully) temporary fix for issue #1383.
+	 * 
+	 *    https://github.com/mono/SkiaSharp/issues/1383
+	 * 
+	 * On Windows, .NET locks are alertable when using the STA threading model and can 
+	 * cause the Windows message loop to be dispatched (typically on WM_PAINT messages. 
+	 * This can lead to re-entrancy and a deadlock on the HandleDictionary lock. 
+	 * 
+	 * This fix replaces the ReaderWriteLockSlim instance on Windows with a native Win32 
+	 * CRITICAL_SECTION.
+	 */
+
+	/// <summary>
+	/// Abstracts a platform dependant lock implementation
+	/// </summary>
+	interface IPlatformLock
+	{
+		void EnterReadLock ();
+		void ExitReadLock ();
+		void EnterWriteLock ();
+		void ExitWriteLock ();
+		void EnterUpgradeableReadLock ();
+		void ExitUpgradeableReadLock ();
+	}
+
+	/// <summary>
+	/// Helper class to create a IPlatformLock instance depending on the platform
+	/// </summary>
+	static class PlatformLock
+	{
+		public static IPlatformLock Create ()
+		{
+			if (RuntimeInformation.IsOSPlatform (OSPlatform.Windows))
+				return new NonAlertableWin32Lock ();
+			else
+				return new ReadWriteLock ();
+		}
+	}
+
+	/// <summary>
+	/// Non-Windows platform lock uses ReaderWriteLockSlim
+	/// </summary>
+	class ReadWriteLock : IPlatformLock
+	{
+		public void EnterReadLock () => _lock.EnterReadLock ();
+		public void ExitReadLock () => _lock.ExitReadLock ();
+		public void EnterWriteLock () => _lock.EnterWriteLock ();
+		public void ExitWriteLock () => _lock.ExitWriteLock ();
+		public void EnterUpgradeableReadLock () => _lock.EnterUpgradeableReadLock ();
+		public void ExitUpgradeableReadLock () => _lock.ExitUpgradeableReadLock ();
+
+		ReaderWriterLockSlim _lock = new ReaderWriterLockSlim ();
+	}
+
+	/// <summary>
+	/// Windows platform lock uses Win32 CRITICAL_SECTION
+	/// </summary>
+	class NonAlertableWin32Lock : IPlatformLock
+	{
+		public NonAlertableWin32Lock ()
+		{
+			_cs = Marshal.AllocHGlobal (Marshal.SizeOf<CRITICAL_SECTION> ());
+			InitializeCriticalSectionEx (_cs, 4000, 0);
```

Comment:
Should these calls check the result and throw using `Marshal.GetLastWin32Error`? At least we can catch an error and not break later on in a random place.

Reply:
I've put in a check on the AllocHGlobal to throw an out of memory exception.  I'm not sure what conditions InitCritSec would fail (I've never seen one fail) so not sure what exception to throw.

Reply:
I mean the InitializeCriticalSectionEx may return false and have an error.

Reply:
Yes, I understand that... I just wasn't sure what exception to throw? Suggestions?

Reply:
Looking at people's code online, seems that they just use void and thus never care. I wonder why. Maybe it never fails?

Reply:
I looked in CoreRT and Sqlite. So I guess just ignore for now... I saw a comment on `InitializeCriticalSection` and `InitializeCriticalSectionAndSpinCount` so maybe it also applies:

> This function always succeeds and returns a nonzero value.
>
> Windows Server 2003 and Windows XP:  If the function succeeds, the return value is nonzero. If the function fails, the return value is zero (0). To get extended error information, call GetLastError. Starting with Windows Vista, the InitializeCriticalSectionAndSpinCount function always succeeds, even in low memory situations.

Reply:
Not 100% sure, but I suspect that InitializeCriticalSection just initializes the fields of the CRITICAL_SECTION structure... so it could be that it only fails if you pass it bad parameters like a null pointer for example.